In [13]:
import math
import time

#the different trials we are using


#Calculate naive summation
def naive_summation(n):
    N=10**n
    nsum=0
    for i in range(0,N):
        nsum +=math.e
    return nsum

#Calculate compensated summation
def compensated_summation(n):
    N=10**n
    csum=0
    e=0
    for i in range(0,N):
        temp=csum
        y=math.e+e
        csum=temp+y
        e=(temp-csum)+y
    csum+=e
    return csum

#calculates the time and summation->then prints
def naive_summation_time(n):
    start_time = time.time()
    nsum=naive_summation(n)#calculates the naive sum
    end_time = time.time()
    elasped_time = end_time-start_time
    print('Naive Summation of N^'+str(n)+' is: '+str(nsum)+'. Time taken is: '+str(elasped_time))#prints everythng
    return nsum

#calculates the time and summation->then prints
def compensated_summation_time(n):
    start_time = time.time() #starts time
    csum=compensated_summation(n) #calculates the compensated sum
    end_time = time.time() # end time
    elasped_time = end_time-start_time #calculate the elasped time
    print('Compensated Summation of N^'+str(n)+' is: '+str(csum)+'. Time taken is: '+str(elasped_time)+" seconds.")#prints everythng
    return csum
    
#Calculate the Exact value
def exact_value(n):
    N=10**n
    result=N*math.e
    return result

#Calculate absolute error 
def absolute_error(approx,exact):
    result= math.fabs(approx - exact) 
    return result

#Calculate relative error
def relative_error(approx,exact):
    result=math.fabs(approx-exact)/math.fabs(exact)
    return result

#calculate the exact sums
esum6=exact_value(6)
esum7=exact_value(7)
esum8=exact_value(8)
esum9=exact_value(9)

#calculate the naive sum
nsum6=naive_summation_time(6)
print('N^6: The Absolute Error is '+str(absolute_error(nsum6,esum6))+' and the Relative Error is '+str(relative_error(nsum6,esum6)))

nsum7 = naive_summation_time(7)
print('N^7: The Absolute Error is '+str(absolute_error(nsum7,esum7))+' and the Relative Error is '+str(relative_error(nsum7,esum7)))

nsum8=naive_summation_time(8)
print('N^8: The Absolute Error is '+str(absolute_error(nsum8,esum8))+' and the Relative Error is '+str(relative_error(nsum8,esum8)))

nsum9=naive_summation_time(9)
print('N^9: The Absolute Error is '+str(absolute_error(nsum9,esum9))+' and the Relative Error is '+str(relative_error(nsum9,esum9)))

#calculate the compensated sum
csum6=compensated_summation_time(6)
print('N^6: The Absolute Error is '+str(absolute_error(csum6,esum6))+' and the Relative Error is '+str(relative_error(csum6,esum6)))

csum7=compensated_summation_time(7)
print('N^7: The Absolute Error is '+str(absolute_error(csum7,esum7))+' and the Relative Error is '+str(relative_error(csum7,esum7)))

csum8=compensated_summation_time(8)
print('N^8: The Absolute Error is '+str(absolute_error(csum8,esum8))+' and the Relative Error is '+str(relative_error(csum8,esum8)))
csum9=compensated_summation_time(9)
print('N^9: The Absolute Error is '+str(absolute_error(csum9,esum9))+' and the Relative Error is '+str(relative_error(csum9,esum9)))


Naive Summation of N^6 is: 2718281.828467976. Time taken is: 0.15502595901489258
N^6: The Absolute Error is 8.930917829275131e-06 and the Relative Error is 3.2855010601818064e-12
Naive Summation of N^7 is: 27182818.280961137. Time taken is: 1.7413990497589111
N^7: The Absolute Error is 0.0036293119192123413 and the Relative Error is 1.335149240676691e-10
Naive Summation of N^8 is: 271828183.31291556. Time taken is: 14.703012943267822
N^8: The Absolute Error is 0.467011034488678 and the Relative Error is 1.7180375838859204e-09
Naive Summation of N^9 is: 2718281783.988457. Time taken is: 200.3285539150238
N^9: The Absolute Error is 44.470587730407715 and the Relative Error is 1.6359814962827992e-08
Compensated Summation of N^6 is: 2718281.828459045. Time taken is: 0.29715800285339355 seconds.
N^6: The Absolute Error is 0.0 and the Relative Error is 0.0
Compensated Summation of N^7 is: 27182818.28459045. Time taken is: 3.1064341068267822 seconds.
N^7: The Absolute Error is 0.0 and the Rel

# |N^n|Naive Summation|Time Taken|Compensenated Summation|Time Taken|Absolute Error-Naive|Relative Error-Naive|Absolute Error-Compensated|Relative Error-Compensated|
|---|---|---|---|---|---|---|---|---|
|10^6|2718281.828467976|0.15502595901489258|2718281.828459045|0.29715800285339355|8.930917829275131e-06|3.2855010601818064e-12|0.0|0.0|
|10^7|27182818.280961137|1.7413990497589111|27182818.28459045|3.1064341068267822|0.0036293119192123413|1.335149240676691e-10|0.0|0.0|
|10^8|271828183.31291556|14.703012943267822|271828182.8459045|37.43103885650635|0.467011034488678|1.7180375838859204e-09|0.0|0.0|
|10^9|2718281783.988457|200.3285539150238|2718281828.459045|345.9846429824829|44.470587730407715|1.6359814962827992e-08|0.0|0.0|